code/crates/test/app/src/app.rs

```diff
@@ -0,0 +1,317 @@
+
+                sleep(Duration::from_millis(200)).await;
+
+                // We can simply respond by telling the engine to start consensus
+                // at the next height, and provide it with the genesis validator set
+                if reply
+                    .send(ConsensusMsg::StartHeight(
+                        start_height,
+                        genesis.validator_set.clone(),
+                    ))
+                    .is_err()
+                {
+                    error!("Failed to send ConsensusReady reply");
+                }
+            }
+
+            // The next message to handle is the `StartRound` message, signaling to the app
+            // that consensus has entered a new round (including the initial round 0)
+            AppMsg::StartedRound {
+                height,
+                round,
+                proposer,
+            } => {
+                info!(%height, %round, %proposer, "Started round");
+
+                // We can use that opportunity to update our internal state
+                state.current_height = height;
+                state.current_round = round;
+                state.current_proposer = Some(proposer);
+            }
+
+            // At some point, we may end up being the proposer for that round, and the engine
+            // will then ask us for a value to propose to the other validators.
+            AppMsg::GetValue {
+                height,
+                round,
+                timeout: _,
+                reply,
+            } => {
+                // NOTE: We can ignore the timeout as we are building the value right away.
+                // If we were let's say reaping as many txes from a mempool and executing them,
+                // then we would need to respect the timeout and stop at a certain point.
+
+                info!(%height, %round, "Consensus is requesting a value to propose");
+
+                // Here it is important that, if we have previously built a value for this height and round,
+                // we send back the very same value. We will not go into details here but this has to do
+                // with crash recovery and is not strictly necessary in this example app since all our state
+                // is kept in-memory and therefore is not crash tolerant at all.
+                if let Some(proposal) = state.get_previously_built_value(height, round).await? {
+                    info!(value = %proposal.value.id(), "Re-using previously built value");
+
+                    if reply.send(proposal).is_err() {
+                        error!("Failed to send GetValue reply");
+                    }
+
+                    return Ok(());
+                }
+
+                // If we have not previously built a value for that very same height and round,
+                // we need to create a new value to propose and send it back to consensus.
+                let proposal = state.propose_value(height, round).await?;
```

Comment:
The way we build the block/ parts vs propose value is strange. The normal order should be:
- create block
- stream parts
- call `propose_value()` and determine the proposal based on the block created before
I understand this app constructs the parts based on the primes of the randomly generated value (u64) but maybe we can hide that deeper.